### PR TITLE
Add Provider Engine Support

### PR DESCRIPTION
**Pull request by OR13**

### CODE REVIEW

File: src/Toolbox/Toolbox.spec.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            let addr = Toolbox.getDefaultAddressFromWallet(wallet);
+        })
+    })
+
+    describe('supports light wallet sign/recover', () => {
+
+        it('when initialized with wallet', async () => {
+            var mnemonic = "couch solve unique spirit wine fine occur rhythm foot feature glory away";
+            let wallet = Toolbox.getWalletFromMnemonic(default_mnemonic);
+            TransmuteFramework.init({
```
Comment (eolszewski) [inline src/Toolbox/Toolbox.spec.ts]: @OR13 where is this used in this test?

File: src/Toolbox/Toolbox.spec.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    let default_mnemonic = 'shoulder biology glory jacket pony circle nuclear wood arrow erupt stomach sing';
+
+    before(async () => {
+        account_addresses = await TransmuteFramework.getAccounts();
+        account = account_addresses[0];
+    })
+
+    describe('.sign', () => {
+        it('returns a signature', async () => {
+            let { web3 } = TransmuteFramework;
```
Comment (eolszewski) [inline src/Toolbox/Toolbox.spec.ts]: Why are we destructuring this?

File: src/Toolbox/Toolbox.spec.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            TransmuteFramework.init({
+                env: 'infura',
+                wallet: wallet,
+                aca: accessControlArtifacts,
+                esa: eventStoreArtifacts,
+                esfa: eventStoreFactoryArtifacts
+            })
+            let addr = Toolbox.getDefaultAddressFromWallet(wallet);
+            let signature = await Toolbox.sign(addr, 'hello');
+            let recoverd_addr = await Toolbox.recover(addr, 'hello', signature);
```
Comment (eolszewski) [inline src/Toolbox/Toolbox.spec.ts]: recovered_addr*

File: src/Toolbox/Toolbox.spec.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            assert(recoverd_addr === addr)
+        })
+
+        it('when wallet added later...', async () => {
+            var mnemonic = Toolbox.generateMnemonic();
+            let wallet = Toolbox.getWalletFromMnemonic(mnemonic);
+            TransmuteFramework.init()
+            let confg = TransmuteFramework.config;
```
Comment (eolszewski) [inline src/Toolbox/Toolbox.spec.ts]: config*

File: src/Toolbox/Toolbox.spec.ts (hunk 2 above)
Comment (eolszewski) [inline src/Toolbox/Toolbox.spec.ts]: We already get it from the top of the file, no?

File: src/Toolbox/Toolbox.spec.ts (hunk 1 above)
Comment (eolszewski) [inline src/Toolbox/Toolbox.spec.ts]: Do we get a new Toolbox when we do this?
I guess so

File: src/Toolbox/Toolbox.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+export class Toolbox {
+
+  constructor(
+    public framework: ITransmuteFramework,
+  ) {
+
+  }
+
+  public sign = (address, message) => {
+    return new Promise((resolve, reject) => {
```
Comment (eolszewski) [inline src/Toolbox/Toolbox.ts]: async / await can't be used here?

File: src/Toolbox/Toolbox.spec.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
+'use strict'
+
+import TransmuteFramework from '../TransmuteFramework'
+
+const { web3, Toolbox, } = TransmuteFramework.init()
```
Comment (eolszewski) [inline src/Toolbox/Toolbox.spec.ts]: extra comma

File: src/Toolbox/Toolbox.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        // console.log(message.length)
+        // THIS IS LAME... testrpc... I hate you....
+        this.framework.web3.eth.sign(address, message, (err, signature) => {
+            if (err) {
```
Comment (eolszewski) [inline src/Toolbox/Toolbox.ts]: what err would occur here?

Comment (eolszewski) [inline src/Toolbox/Toolbox.ts]: why does hashing this fix the err?

File: src/Toolbox/Toolbox.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    return "0x" + wallet.getAddress().toString("hex");
+  }
+
+  public getDefaultAddressFromMnemonic = (mnemonic: string) => {
+    return this.getDefaultAddressFromWallet(this.getWalletFromMnemonic(mnemonic))
+  }
+
+
+}
+
```
Comment (eolszewski) [inline src/Toolbox/Toolbox.ts]: extra space

File: src/TransmuteFramework.spec.ts [hunk 9]
```diff
@@ -17,20 +24,35 @@ describe('TransmuteFramework', () => {
...
-            // console.log(TransmuteFramework.web3)
-            assert(TransmuteFramework.web3._requestManager.provider.host === 'http://localhost:8545')
+            // TODO: add tests here...
         })
         it('should support infura ropsten', async () => {
-            const accessControlArtifacts = require('../build/contracts/RBAC')
-            const eventStoreArtifacts = require('../build/contracts/RBACEventStore')
-            const eventStoreFactoryArtifacts = require('../build/contracts/RBACEventStoreFactory')
+
+            TransmuteFramework.init({
```
Comment (eolszewski) [inline src/TransmuteFramework.spec.ts]: Should we initialize this with a wallet? Just curious.

Comment (eolszewski) [inline src/TransmuteFramework.spec.ts]: Nevermind, I see the initialization below.

File: src/TransmuteFramework.spec.ts [hunk 10]
```diff
@@ -17,20 +24,35 @@ describe('TransmuteFramework', () => {
...
+                esa: eventStoreArtifacts,
+                esfa: eventStoreFactoryArtifacts
+            })
+            // TODO: add tests here...
+        })
+
+        it('should support HD Lightwallets', async () => {
+
+            var mnemonic = "couch solve unique spirit wine fine occur rhythm foot feature glory away";
+            var hdwallet = hdkey.fromMasterSeed(bip39.mnemonicToSeed(mnemonic));
```
Comment (eolszewski) [inline src/TransmuteFramework.spec.ts]: Reference - https://github.com/ethereumjs/ethereumjs-wallet/blob/master/hdkey.js

File: src/TransmuteFramework.ts [hunk 11]
```diff
@@ -20,13 +19,19 @@ const eventStoreArtifacts = require('../build/contracts/RBACEventStore')
...
+
+
 
 export interface ITransmuteFrameworkConfig {
   env: string
   aca: any,
   esa: any,
   esfa: any,
-  ipfsConfig?: any
+  ipfsConfig?: any,
```
Comment (eolszewski) [inline src/TransmuteFramework.ts]: I see, these are optional params.

File: src/TransmuteFramework.ts [hunk 12]
```diff
@@ -20,13 +19,19 @@ const eventStoreArtifacts = require('../build/contracts/RBACEventStore')
 const eventStoreFactoryArtifacts = require('../build/contracts/RBACEventStoreFactory')
 
 
+var ProviderEngine = require("web3-provider-engine");
+var WalletSubprovider = require('web3-provider-engine/subproviders/wallet.js');
+var Web3Subprovider = require("web3-provider-engine/subproviders/web3.js");
```
Comment (eolszewski) [inline src/TransmuteFramework.ts]: 👍

File: src/Toolbox/Toolbox.ts (hunk 5 above)
Comment (OR13) [inline src/Toolbox/Toolbox.ts]: I cannot

File: src/Toolbox/Toolbox.ts (hunk 7 above)
Comment (OR13) [inline src/Toolbox/Toolbox.ts]: the way eth.sign works depends on the provider, node and wallet... sometimes only sha3 length messages can be signed.

File: src/Toolbox/Toolbox.spec.ts (hunk 2 above)
Comment (OR13) [inline src/Toolbox/Toolbox.spec.ts]: good call

File: src/Toolbox/Toolbox.spec.ts (hunk 1 above)
Comment (OR13) [inline src/Toolbox/Toolbox.spec.ts]: thats right. The point of this test is to show that you can call init again, and pass a wallet, and use the wallet.

**Pull request merged**